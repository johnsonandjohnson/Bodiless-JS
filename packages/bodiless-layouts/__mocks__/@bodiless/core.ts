const editContext = {
  activate: jest.fn(),
  isEdit: true,
};

export const useEditContext = jest.fn(() => editContext);

const activateOnEffect = {
  setId: jest.fn(),
};
export const useActivateOnEffect = () => activateOnEffect;

export const useNode = jest.fn();

const contextMenuFormInner = jest.fn();
export const contextMenuForm = jest.fn(() => contextMenuFormInner);
