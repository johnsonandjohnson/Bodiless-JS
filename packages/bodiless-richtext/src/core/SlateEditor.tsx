/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  useRef,
  useState,
  useMemo,
  Fragment,
  ComponentType,
} from 'react';
import { Editor, Value } from 'slate';
import { Editor as ReactEditor, EditorProps, Plugin } from 'slate-react';
// @ts-ignore
import PlaceholderPlugin from 'slate-react-placeholder';
import SlateEditorContext from './SlateEditorContext';
import '@material/react-material-icon/dist/material-icon.css';
import { EditorOnChange } from '../Type';

type Props = {
  initialValue: object;
} & EditorProps;

const withSlateEditor = <P extends object> (Component:ComponentType<P>) => (props:P & Props) => {
  const {
    initialValue, value, onChange, placeholder, plugins = [], ...rest
  } = props;
  // A reference to 'slate-react/Content' instance that allows to manipulate the editor
  const editorRef = useRef<ReactEditor>(null);

  // It is important to keep track of internal activeValue
  // state in case outer activeValue is not provided.
  // Value is used in plugins and buttons before Content is mounted and its activeValue is obtained.
  const [localValueState, setLocalValue] = useState<Value>(
    value
    || Value.fromJSON(initialValue),
  );
  const internalOnChange: EditorOnChange = change => {
    const { value: valueFromChange } = change;
    if (typeof onChange === 'function') {
      onChange(change);
    } else if (typeof value !== 'undefined') {
      setLocalValue(valueFromChange);
    }

    return change;
  };

  const pluginWithPlaceholder = (placeholder) ? useMemo(() => [...plugins,
    [
      {
        queries: {
          isEmpty: (editor:Editor) => editor.value.document.text === '',
        },
      },
      PlaceholderPlugin({
        placeholder,
        when: 'isEmpty',
        style: {
          width: '100%',
          display: 'inline',
          height: '0',
          overflow: 'visible',
        },
      }),
    ],
  ], [plugins]) as Plugin[] : plugins;

  const editorContextValue = {
    editorRef,
    // for some reason current.controller doesn't return what it's supposed to
    // but editorRef.current!.controller.controller does.
    editor:
      editorRef.current!
      // @ts-ignore
      && (editorRef.current!.controller.controller as Editor),
    value: value || localValueState,
    editorProps: {
      ...rest,
      plugins: pluginWithPlaceholder,
      onChange: internalOnChange,
      value: value || localValueState,
    },
  };

  return (
    <SlateEditorContext.Provider value={editorContextValue}>
      <Component {...rest as P} />
    </SlateEditorContext.Provider>
  );
};
const SlateEditor = withSlateEditor(Fragment);

export default SlateEditor;
export { withSlateEditor };
