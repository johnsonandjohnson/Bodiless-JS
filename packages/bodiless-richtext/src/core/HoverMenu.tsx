/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React,
{
  useEffect, ComponentType, HTMLProps,
} from 'react';
import ReactDOM from 'react-dom';
import { useEditContext } from '@bodiless/core';
import { useSlateContext } from './SlateEditorContext';
import { EditorContext } from '../Type';

const defaultUI = {
  Menu: 'div',
};

export type UI = {
  Menu?: ComponentType<HTMLProps<HTMLDivElement>> | string;
};

export const getUI = (ui: UI = {}) => ({
  ...defaultUI,
  ...ui,
});

/**
 * Update the menu's absolute position.
 */
function updateMenu(menu: HTMLElement | null, editorContext: EditorContext) {
  const native = window.getSelection();

  if (!menu || !editorContext || !native) return;
  const { value } = editorContext;
  const { fragment, selection } = value;

  /**
   * According to https://github.com/ianstormtaylor/slate/issues/2432
   * we need to detect if the active editor looses the focus (isBlurred equal to true)
   */
  if (selection.isBlurred || selection.isCollapsed || fragment.text === '') {
    menu.removeAttribute('style');
    return;
  }
  const range = native.getRangeAt(0);
  const rect = range.getBoundingClientRect();
  const offsetLeft = rect.left + window.pageXOffset - menu.offsetWidth / 2 + rect.width / 2;

  const { style } = menu;
  style.opacity = '1';
  style.visibility = 'visible';
  style.top = `${rect.top + window.pageYOffset - menu.offsetHeight}px`;
  style.left = `${offsetLeft < 0 ? 15 : offsetLeft}px`;
}

export type HoverMenuProps = {
  children?: any;
  className?: string;
  ui?: UI;
};

const HoverMenu = (props: HoverMenuProps) => {
  const isEditMode = useEditContext().isEdit || null;
  const editorContext: EditorContext = useSlateContext();

  const { ui } = props;
  const { Menu } = getUI(ui);

  const { children, className, ...rest } = props;
  const root = typeof window !== 'undefined' ? window.document.body : null;
  const elementID = `hover-menu-${Math.random().toString(16).slice(2)}`;

  useEffect(() => {
    const element = document.getElementById(elementID);
    updateMenu(element, editorContext);
    return () => {};
  });
  return (
    root
    && isEditMode
    && ReactDOM.createPortal(
      <Menu {...rest} id={elementID} className={className}>
        {children}
      </Menu>,
      root,
    )
  );
};

HoverMenu.displayName = 'HoverMenu';

export default HoverMenu;
