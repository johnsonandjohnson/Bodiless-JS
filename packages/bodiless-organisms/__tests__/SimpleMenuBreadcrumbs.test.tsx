/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mount, ReactWrapper } from 'enzyme';
// eslint-disable-next-line import/no-extraneous-dependencies
import cheerio from 'cheerio';
import { withDefaultContent, withSidecarNodes, ifToggledOn } from '@bodiless/core';
import {
  asBodilessLink, asEditable, withBreadcrumbStartingTrail, Breadcrumbs, withBreadcrumbStore,
} from '@bodiless/components';
import {
  replaceWith,
  withDesign,
  addProps,
} from '@bodiless/fclasses';
import { flowRight } from 'lodash';
import type { BreadcrumbStoreItemsReducer } from '@bodiless/components';

import {
  asMenuBase,
  asBreadcrumbSource,
} from '../src/components/Menu/SimpleMenu';

const { DefaultContentNode } = require('@bodiless/core');

const setPagePath = (pagePath: string) => {
  Object.defineProperty(DefaultContentNode.prototype, 'pagePath', {
    value: pagePath,
    writable: false,
  });
};

const createBreadcrumbComponent = ({
  content = {},
}) => {
  const Source = flowRight(
    asBreadcrumbSource({
      linkNodeKey: 'title$link',
      titleNodeKey: 'title$text',
    }),
    asMenuBase('testMenu'),
  )('ul');

  const BreadcrumbComponent = (props: any) => (
    <>
      <Source />
      <Breadcrumbs {...props} />
    </>
  );

  return flowRight(
    withDefaultContent(content),
    withDesign({
      // @ts-ignore
      BreadcrumbLink: replaceWith(withSidecarNodes(
        asBodilessLink(),
      ))('a'),
      BreadcrumbTitle: replaceWith(
        asEditable()(React.Fragment),
      ),
    }),
    withBreadcrumbStore,
  )(BreadcrumbComponent);
};

// Helper to get the html of only the breadcrumbs.
const breadcrumbHtml = (wrapper: ReactWrapper) => {
  // We use cheerio directly here bc using enzyme's 'render'
  // method generates warnings about not using layout
  // effects on the server.
  const $ = cheerio.load(wrapper.html());
  return $.html($('body>ul').last());
};

const generate2LevelMenuContent = () => ({
  testMenu: {
    items: [
      'home',
      'products',
      'articles',
    ],
  },
  testMenu$home$title$link: {
    href: '/',
  },
  testMenu$home$title$text: {
    text: 'home',
  },
  testMenu$products$title$link: {
    href: '/products',
  },
  testMenu$products$title$text: {
    text: 'products',
  },
  testMenu$products$sublist: {
    items: [
      'productA',
      'productB',
    ],
  },
  'testMenu$products$cham-sublist': {
    component: 'SubMenu',
  },
  testMenu$products$sublist$productA$title$link: {
    href: '/products/productA',
  },
  testMenu$products$sublist$productA$title$text: {
    text: 'productA',
  },
  testMenu$products$sublist$productB$title$link: {
    href: '/products/productB',
  },
  testMenu$products$sublist$productB$title$text: {
    text: 'productB',
  },
  testMenu$articles$title$link: {
    href: '/articles',
  },
  testMenu$articles$title$text: {
    text: 'articles',
  },
  testMenu$articles$sublist: {
    items: [
      'articleA',
    ],
  },
  testMenu$articles$sublist$articleA$title$link: {
    href: '/articles/articleA',
  },
  testMenu$articles$sublist$articleA$title$text: {
    text: 'articleA',
  },
});

describe('asBreadcrumbsClean', () => {
  it('creates breadcrumbs for basic 1-level menu', () => {
    setPagePath('/products');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
            'products',
            'articles',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$home$title$text: {
          text: 'home',
        },
        testMenu$products$title$link: {
          href: '/products',
        },
        testMenu$products$title$text: {
          text: 'products',
        },
        testMenu$articles$title$link: {
          href: '/articles',
        },
        testMenu$articles$title$text: {
          text: 'articles',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('creates breadcrumbs for basic 2-level menu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: generate2LevelMenuContent(),
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it(`when starting trail enabled and
      when menu trail starts with frontpage
      then it does not render home link twice by default`,
  () => {
    setPagePath('/products');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$home$title$text: {
          text: 'home',
        },
        'testMenu$home$cham-sublist': {
          component: 'SubMenu',
        },
        testMenu$home$sublist$productA$title$link: {
          href: '/products',
        },
        testMenu$home$sublist$productA$title$title: {
          href: '/products',
        },
      },
    });
    const CustomBreadcrumb = flowRight(
      withDesign({
        StartingTrail: replaceWith(() => <span>home</span>),
      }),
      withBreadcrumbStartingTrail,
    )(Breadcrumb);
    const wrapper = mount(<CustomBreadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it(`allows override default items reducer
      so that do not render the first item from menu trail`,
  () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: generate2LevelMenuContent(),
    });
    const customReducer: BreadcrumbStoreItemsReducer = items => items
      .filter(item => !item.isFirst())
      .map(item => item.uuid);
    const wrapper = mount(<Breadcrumb itemsReducer={customReducer} />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('renders last menu trail item as link when current page item does not exist in store', () => {
    setPagePath('/products/nonExistingProduct');
    const Breadcrumb = createBreadcrumbComponent({
      content: generate2LevelMenuContent(),
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('allows to force rendering last breadcrumb item as link', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: generate2LevelMenuContent(),
    });
    const wrapper = mount(<Breadcrumb renderLastItemWithoutLink={false} />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('allows styling current page item derived from menu', () => {
    setPagePath('/products/productA');
    const BaseBreadcrumbs = createBreadcrumbComponent({
      content: generate2LevelMenuContent(),
    });
    const TestBreadcrumbs = flowRight(
      withDesign({
        BreadcrumbItem: ifToggledOn(
          ({ isCurrentPage }: any) => isCurrentPage,
        )(addProps({ className: 'font-bold' })),
      }),
    )(BaseBreadcrumbs);
    const wrapper = mount(<TestBreadcrumbs />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('does not apply current page styles to the item which is the last derived from the menu but not current page path', () => {
    setPagePath('/products/nonExistingProduct');
    const BaseBreadcrumbs = createBreadcrumbComponent({
      content: generate2LevelMenuContent(),
    });
    const TestBreadcrumbs = flowRight(
      withDesign({
        BreadcrumbItem: ifToggledOn(
          ({ isCurrentPage }: any) => isCurrentPage,
        )(addProps({ className: 'font-bold' })),
      }),
    )(BaseBreadcrumbs);
    const wrapper = mount(<TestBreadcrumbs />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
});
