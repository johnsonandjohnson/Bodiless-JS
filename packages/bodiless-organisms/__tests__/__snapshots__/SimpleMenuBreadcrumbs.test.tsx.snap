// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asBreadcrumbsClean allows override default items reducer
      so that do not render the first item from menu trail 1`] = `"<ul><li><span>productA</span></li></ul>"`;

exports[`asBreadcrumbsClean allows styling current page item derived from menu 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li class=\\"font-bold\\"><span>productA</span></li></ul>"`;

exports[`asBreadcrumbsClean allows to force rendering last breadcrumb item as link 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li><a href=\\"/products/productA/\\"><span>productA</span></a></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for basic 1-level menu 1`] = `"<ul><li><span>products</span></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for basic 2-level menu 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li><span>productA</span></li></ul>"`;

exports[`asBreadcrumbsClean does not apply current page styles to the item which is the last derived from the menu but not current page path 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li></ul>"`;

exports[`asBreadcrumbsClean renders last menu trail item as link when current page item does not exist in store 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li></ul>"`;

exports[`asBreadcrumbsClean when starting trail enabled and
      when menu trail starts with frontpage
      then it does not render home link twice by default 1`] = `"<ul><li><span>home</span></li></ul>"`;
